### ISSUE TITLE
trainer evaluation and the hand-made evaluation results are different

### ISSUE THREAD

**Issue by min6434** [Microsoft/CNTK#1768]
I used ImageDeserializer as a MinibatchSource and added mean xml file.
Using the same test data evaluation shows 1.6% however the handmade evaluation function shows 4.69%.

I think input feature normalization must be different but don't know the precise implementation inside of the MinibatchSource .

Here are the source code below.

```
#####################################################################################################################################################
# Define the reader for both training and evaluation action.
def create_reader(map_file, mean_file, image_width, image_height, num_channels, num_classes, train):
    "Image reader creation for the learner"
    if not os.path.exists(map_file) or not os.path.exists(mean_file):
        raise RuntimeError("This tutorials depends 201A tutorials, please run 201A first.")

    # transformation pipeline for the features has jitter/crop only when training
    transforms = []
   # if train:
   #     transforms += [
   #         xforms.crop(crop_type='randomside', side_ratio=0.8) # train uses data augmentation (translation only)
   #     ]
    transforms += [
        xforms.scale(width=image_width, height=image_height, channels=num_channels, interpolations='linear'),
        xforms.mean(mean_file)
    ]
    # deserializer
    return MinibatchSource(ImageDeserializer(map_file, StreamDefs(
        features = StreamDef(field='image', transforms=transforms), # first column in map file is referred to as 'image'
        labels   = StreamDef(field='label', shape=num_classes)      # and second as 'label'
    )))

def create_basic_model_with_batch_normalization(input, out_dims):

    with default_options(activation=relu):
        model = Sequential([
            For(range(3), lambda i: [
                Convolution((5,5), [32,32,64][i], init=glorot_uniform(), pad=True),
                BatchNormalization(map_rank=1),
                MaxPooling((3,3), strides=(2,2))
            ]),
            Dense(64, init=glorot_uniform()),
            Dropout(0.25),
            BatchNormalization(map_rank=1),
            Dense(out_dims, init=glorot_uniform(), activation=None)
        ])

    return model(input)

#####################################################################################################################################################
def train_and_evaluate(reader_train, reader_test, image_width, image_height, num_channels, num_classes, max_epochs, model_func):
    # Input variables denoting the features and label data
    input_var = C.input((num_channels, image_height, image_width))
    label_var = C.input((num_classes))
    
    # Normalize the input
    feature_scale = 1.0 / 256.0
    input_var_norm = element_times(feature_scale, input_var)
    
    # apply model to input
    z = model_func(input_var_norm, out_dims=2)

    #
    # Training action
    #

    # loss and metric
    ce = cross_entropy_with_softmax(z, label_var)
    pe = classification_error(z, label_var)

    # training config
    epoch_size     =  461
    minibatch_size = 64

    # Set training parameters
    lr_per_minibatch       = learning_rate_schedule([0.01]*10 + [0.003]*10 + [0.001], UnitType.minibatch, epoch_size)
    momentum_time_constant = momentum_as_time_constant_schedule(-minibatch_size/np.log(0.9))
    l2_reg_weight          = 0.001
    
    # trainer object
    learner = momentum_sgd(z.parameters, 
                           lr = lr_per_minibatch, momentum = momentum_time_constant, 
                           l2_regularization_weight=l2_reg_weight)
    progress_printer = ProgressPrinter(tag='Training', num_epochs=max_epochs)
    trainer = Trainer(z, (ce, pe), [learner], [progress_printer])

    # define mapping from reader streams to network inputs
    input_map = {
        input_var: reader_train.streams.features,
        label_var: reader_train.streams.labels
    }

    log_number_of_parameters(z) ; print()

    # perform model training
    batch_index = 0
    plot_data = {'batchindex':[], 'loss':[], 'error':[]}
    for epoch in range(max_epochs):       # loop over epochs
        sample_count = 0
        while sample_count < epoch_size:  # loop over minibatches in the epoch
            data = reader_train.next_minibatch(min(minibatch_size, epoch_size - sample_count), input_map=input_map) # fetch minibatch.
            trainer.train_minibatch(data)                                   # update model with it

            sample_count += data[label_var].num_samples                     # count samples processed so far
            
            # For visualization...            
            plot_data['batchindex'].append(batch_index)
            plot_data['loss'].append(trainer.previous_minibatch_loss_average)
            plot_data['error'].append(trainer.previous_minibatch_evaluation_average)
            
            batch_index += 1
        trainer.summarize_training_progress()
        
    #
    # Evaluation action
    #
    epoch_size     = 64
    minibatch_size = 64

    # process minibatches and evaluate the model
    metric_numer    = 0
    metric_denom    = 0
    sample_count    = 0
    minibatch_index = 0

    while sample_count < epoch_size:
        current_minibatch = min(minibatch_size, epoch_size - sample_count)

        # Fetch next test min batch.
        data = reader_test.next_minibatch(current_minibatch, input_map=input_map)

        # minibatch data to be trained with
        metric_numer += trainer.test_minibatch(data) * current_minibatch
        metric_denom += current_minibatch

        # Keep track of the number of samples processed so far.
        sample_count += data[label_var].num_samples
        minibatch_index += 1

    print("")
    print("Final Results: Minibatch[1-{}]: errs = {:0.1f}% * {}".format(minibatch_index+1, (metric_numer*100.0)/metric_denom, metric_denom))
    print("")
    
    # Visualize training result:
    window_width            = 16
    loss_cumsum             = np.cumsum(np.insert(plot_data['loss'], 0, 0)) 
    error_cumsum            = np.cumsum(np.insert(plot_data['error'], 0, 0)) 

    # Moving average.
    plot_data['batchindex'] = np.insert(plot_data['batchindex'], 0, 0)[window_width:]
    plot_data['avg_loss']   = (loss_cumsum[window_width:] - loss_cumsum[:-window_width]) / window_width
    plot_data['avg_error']  = (error_cumsum[window_width:] - error_cumsum[:-window_width]) / window_width
    
    plt.figure(1)
    plt.subplot(211)
    plt.plot(plot_data["batchindex"], plot_data["avg_loss"], 'b--')
    plt.xlabel('Minibatch number')
    plt.ylabel('Loss')
    plt.title('Minibatch run vs. Training loss ')

    plt.subplot(212)
    plt.plot(plot_data["batchindex"], plot_data["avg_error"], 'r--')
    plt.xlabel('Minibatch number')
    plt.ylabel('Label Prediction Error')
    plt.title('Minibatch run vs. Label Prediction Error ')
    plt.tight_layout()
    plt.show()
    
    return softmax(z)


import PIL
#####################################################################################################################################################
def eval(pred_op, image_path, image_mean):
    label_lookup = ["healty tissue", "metastases"]
    image_data   = np.array(PIL.Image.open(image_path), dtype=np.float32)
    
    if image_data.shape[0] != 256 or image_data.shape[1] != 256:
        image_data = imresize(image_data, (256, 256))
    
    image_data = image_data.astype(dtype = np.float32)
    image_data   = np.ascontiguousarray(np.transpose(image_data, (2, 0, 1)))
    image_data  -= image_mean
    result       = np.squeeze(pred_op.eval({pred_op.arguments[0]:[image_data]}))
    
    # Return top 3 results:
    top_count = 2
    result_indices = (-np.array(result)).argsort()[:top_count]

    #print("Top 2 predictions:")
    #for i in range(top_count):
    #    print("\tLabel: {:10s}, confidence: {:.2f}%".format(label_lookup[result_indices[i]], result[result_indices[i]] * 100))
    
    return result_indices[0]

# Change map text files into the CNTK format
changeCvediaToCNTKmap(os.path.join(DataPath,'train.txt'), os.path.join(DataPath,'train_cntk.txt'), DataPath)
changeCvediaToCNTKmap(os.path.join(DataPath,'validate.txt'), os.path.join(DataPath,'validate_cntk.txt'), DataPath)

# Calculate average pixel data and put them into the XML for CNTK
meanImg = saveMean(os.path.join(DataPath,'train_cntk.txt'))
saveMeanXML(os.path.join(DataPath,'breast_mean.xml'), meanImg, ImagSize)

# Create image readers
reader_train = create_reader(os.path.join(DataPath,'train_cntk.txt'), os.path.join(DataPath,'breast_mean.xml'), image_width, image_height, num_channels, num_classes, True)
reader_test  = create_reader(os.path.join(DataPath,'validate_cntk.txt'), os.path.join(DataPath,'breast_mean.xml'), image_width, image_height, num_channels, num_classes, False)

pred_basic_model_bn = train_and_evaluate(reader_train, reader_test, image_width, image_height, num_channels, num_classes, max_epochs=100,\
    model_func=create_basic_model_with_batch_normalization)

label_lookup = ["healty tissue", "metastases"]
nTotal = 0
nFalse = 0
for i, line in enumerate(open("""D:\\Source\\Repos\\CNTK_sources\\CNTK_Breast\\SygDNXzjqQxPAWC2A7Pes3L2m9EBY2dJ\\validate_cntk.txt""", 'r')):
    imgFile, label = line.split('\t')
    result = eval(pred_basic_model_bn, imgFile, meanImg)
    nTotal += 1
    if result != int(label):
        print("real value: ", label_lookup[int(label)], end = ", ")
        print("network result: ", label_lookup[result])
        nFalse += 1

print(nFalse/nTotal*100)
```

**Comment by min6434** [comment]
I think the answer to the problem could be this.
https://github.com/Microsoft/CNTK/issues/1698
However even though I used opencv to open image to get BRG image the result doesn't seem to change.
```
def saveMean(inputFile, image_height, image_width, num_channels, num_samples):
    "Save mean value of the train images"
    meanImg = np.zeros((num_channels,image_width,image_height), dtype = np.float32)
    
    for line in open(inputFile, 'r'):
        imgFile, label = line.split('\t')
        img = cv.imread(imgFile)
        if img.shape[0] != image_width or img.shape[1] != image_height:
            img = imresize(img, (image_width, image_height))

        img = np.array(img, dtype = np.float32)
        img = np.swapaxes(img,0,2)
        img = np.swapaxes(img,1,2)
        meanImg += img
        
    meanImg = np.divide(meanImg, num_samples)
    return meanImg

def eval(pred_op, image_path, image_mean):
    label_lookup = ["healty tissue", "metastases"]
    image_data = cv.imread(image_path)
    
    if image_data.shape[0] != 256 or image_data.shape[1] != 256:
        image_data = imresize(image_data, (256, 256))
    
    image_data = image_data.astype(dtype = np.float32)
    image_data = np.ascontiguousarray(np.transpose(image_data, (2, 0, 1)))

    image_data  -= image_mean
    image_data = np.ascontiguousarray(image_data)
    result       = np.squeeze(pred_op.eval({pred_op.arguments[0]:[image_data]}))
    
    # Return top 3 results:
    top_count = 2
    result_indices = (-np.array(result)).argsort()[:top_count]
    
    return result_indices[0]
```

**Comment by min6434** [comment]
CNKT Image deserealizer uses (channel, height, width) not (channel, width, height).
Also mean.xml should be normalized by 256 since the input is also normalized in this tutorial.
Surprisingly evaluation result using the cntk.eval function shows wrong accuarcy when I normalize the data using the mean and if I substract nothing then it shows the same result.
I think CNTK had some change so the network do the mean substraction using the mean.xml which is appied at the training and test minibatch.
